### ISSUE TITLE
Store commissions flag in payment meta

### ISSUE THREAD

**Issue by mintplugins** [easydigitaldownloads/EDD-Commissions#140]
**Background:**
The function which records commissions is hooked to the "edd_update_payment_status" and only runs if the "old_status" of the payment is not "publish" or "complete". Otherwise it exits and does nothing. This is to prevent double commission calculations. 

**The Problem**
This doesn't prevent duplicate commissions from being awarded/recorded if the site owner changes the payment status to "pending" and then back to "complete". For a site owner that isn't aware doing that will regenerate/duplicate commissions for that payment, they could end up double paying their commissions.

**Proposed Solution**
Instead of relying on the payment status, I propose we should set a flag in the Payment Meta which is only set to "true" if commissions have been calculated and recorded. 

Then this block of code could be removed:

```
// Check if the payment was already set to complete
	if( $old_status == 'publish' || $old_status == 'complete' )
		return; // Make sure that payments are only completed once

	// Make sure the commission is only recorded when new status is complete
	if( $new_status != 'publish' && $new_status != 'complete' )
		return;

```

and replaced with something like

```
if ( $commissions_already_calculated ){
     return;
}
```

Additionally, I propose we add a second flag which deciphers whether it is the right time to calculate commissions or not. For example, you may want to delay the calculation of commissions based on a customer action - like whether they download the file before a refund period has passed (as a rough example). This could be done with a filter like this:

```
if ( $commissions_already_calculated ){
     return;
}

if ( !apply_filters( 'eddc_calculate_commissions_now', true, $payment_id ) ){
    return;
}
```

**Comment by pippinsplugins** [comment]
Yep, that'll do!

**Comment by pippinsplugins** [comment]
@mintplugins PR submitted.

**Comment by mintplugins** [comment]
@pippinsplugins Can we also remove this?

```
// Check if the payment was already set to complete
    if( $old_status == 'publish' || $old_status == 'complete' )
        return; // Make sure that payments are only completed once

    // Make sure the commission is only recorded when new status is complete
    if( $new_status != 'publish' && $new_status != 'complete' )
        return;
```

**Comment by pippinsplugins** [comment]
@mintplugins We could remove the old status check though it doesn't really benefit us, at least not in any way that I see.

**Comment by mintplugins** [comment]
@pippinsplugins It would allow for commissions to not be generated at other times via add-ons.  For example, you may want to delay the calculation of commissions based on a customer action - like whether they download the file before a refund period has passed (as a rough example).

**Comment by pippinsplugins** [comment]
I'm not following why the status affects the ability to do that. Could you explain?

**Comment by mintplugins** [comment]
If we follow the whole scenario through it would look like this:

1) Customer makes purchase
2) Payment status is set to "Complete"
3) Imagine a filtered function to which delays commissions being generated until 60 days have passed.
4) 60 days pass and now no refunds can be given. Thus, it is now "safe" to generate commissions.
5) The $old_status was "complete" and commissions were not generated yet. In this scenario, unless we remove that if statement, no commissions would be generated.

**Comment by pippinsplugins** [comment]
That all makes sense. In that scenario, however, I don't believe it's right to delay creating the commission. Instead I feel we should have a new status for commissions of `pending`. Commissions should always be created immediately but they can be flagged as ineligible for payout until a later time through a pending status.

That is how we handle it in AffiliateWP and I would propose doing it here.

**Comment by mintplugins** [comment]
I have another potentially-real scenario which **requires** they not be generated immediately. 

If the commission rate is based on user action. 

Imagine the following:
- A single product has multiple files created by different people.
- The customer only downloads file 1 and not file 2. The store wishes to pay its creators based on  which files the customer chose to download. After a time period has passed, an (imaginary) function could check which files were downloaded by the user and generate a commission **rate** based on that user action. This can only be done if commissions are not forced to be generated at the time of sale.

**Comment by mintplugins** [comment]
By default they should still be generated at the time of sale. But removing that code black allows for integrations like the one described above.

**Comment by pippinsplugins** [comment]
Now that's an interesting use case. Going to think on it a bit more.

**Comment by mintplugins** [comment]
Great :) I think commissions would become much more powerful and expandable if we allowed them to be created at times other than directly after sale. This is why I believe it becomes important to add a flag that isn't related to payment status. Let me know your thoughts!

**Comment by mintplugins** [comment]
I think I may have come to the conclusion that this is not worth working towards. 

Here's my thinking: In the scenario I described above, I realized that a single flag for the entire payment is not enough. You would need to have a "Commissions Already Generated" flag for **each item purchased**. This would be because some items purchased would almost definitely need to have their commissions generated at different times. Therefore a single flag for the entire payment is not specific enough.
